doc/reference/ubuntu-image.rst

```diff
@@ -0,0 +1,246 @@
+
+--disk-info DISK-INFO-CONTENTS
+    File to be used as ``.disk/info`` on the root file system of the image.  This file can contain useful information about the target image, such as image identification data, system name, build timestamp, etc.
+
+-c CHANNEL, --channel CHANNEL
+    The default Snap channel to use while preseeding the image.
+
+--sector-size SIZE
+    When creating the disk-image file, use the given sector size.  This can be either 512 or 4096 (4k sector size), defaulting to 512.
+
+
+State-machine options
+~~~~~~~~~~~~~~~~~~~~~
+
+.. caution:: The options described here are primarily for debugging and testing purposes and should not be considered part of the stable, public API.  State-machine step numbers and names can change between releases.
+
+``ubuntu-image`` internally runs a state machine to create the disk image. These are some options for controlling this state machine.  Other than ``--workdir``, these options are mutually exclusive.  When ``--until`` or ``--thru`` is given, the state machine can be resumed later with ``--resume``, but ``--workdir`` must be given in that case since the state is saved in a ``ubuntu-image.json`` file in the working directory.
+
+-w DIRECTORY, --workdir DIRECTORY
+    The working directory in which to download and unpack all the source files for the image.  This directory can exist or not, and it is not removed after this program exits.  If not given, a temporary working directory is used instead, which *is* deleted after this program exits.  Use ``--workdir`` if you want to be able to resume a partial state-machine run.  The ``gadget.yaml`` file is copied to the working directory after it's downloaded.
+
+-u STEP, --until STEP
+    Run the state machine until the given ``STEP``, non-inclusively.  ``STEP`` is the name of a state-machine method. See the `List of steps`_.
+
+-t STEP, --thru STEP
+    Run the state machine until the given ``STEP``, inclusively.  ``STEP`` is the name of a state-machine method. See the `List of steps`_.
+
+-r, --resume
+    Continue the state machine from the previously saved state.  It returns an error if there is no previous state.
+
+
+Files used by ``ubuntu-image``
+------------------------------
+
+* |gadgetyaml|_
+* `model assertion`_
+* `gadget tree`_ (example)
+* `cloud-config`_
+
+
+Environment variables
+---------------------
+
+The following environment variables are recognised by ``ubuntu-image``.
+
+``UBUNTU_IMAGE_PRESERVE_UNPACK``
+    When set, the variable specifies the directory for preserving a pristine copy of the unpacked gadget contents.  The directory must exist, and an ``unpack`` directory will be created under this directory.  The full contents of the ``<workdir>/unpack`` directory after the ``snap prepare-image`` sub-command has run is copied here.
+
+There are a few other environment variables used for building and testing only.
+
+
+.. _list of steps:
+
+List of steps
```

Comment:
I'm removing the list of steps. When the `--dry-run` option is implemented, we can mention it in appropriate places.

Wouldn't it be useful to have an option to list all potential steps for reference?

Reply:
I am not sure about that. Because I do not want users thinking "backward" and trying to hack their image definition to end up executing a specific step. For now we do not really make any promise regarding the order of steps or even the name and existence of steps between versions. This is an advanced feature and we do not expect most users using it. We are thinking about creating "meta-states" representing a high-level state (and regrouping several steps) that we could keep stable between versions, but we are not there yet.

So for now I think only listing steps that will execute for a given configuration is fine.